### DISCUSSION TITLE
[ARCHIVED] Формулировка определения "feature"

### DISCUSSION THREAD

**Discussion by azinit** [feature-sliced/documentation#23] [Archive]
# Дискуссия закрыта, см выводы по #78 
---
---
---

> См. [начало дискуссии](https://github.com/feature-sliced/wiki/discussions/18#discussioncomment-422779)

**Если вкратце - есть ощущение, что мы по-разному понимаем, что есть фича**

## Хотелось бы (в целом):
- **В фичах лежит код, который несёт бизнес ценность.**
- Фича должна принести пользу конечному пользователю, в идеале достаточно маленькую, чтобы можно было компонировать
- Фича отражала часть бизнес-логики
- [Фича была самодостаточной, user-facing, (возможно) реюзабельной, комплексной и несущей определенную логику](https://github.com/feature-sliced/wiki/blob/master/old/feature-driven/FAQ.md#feature-what-is)

## При этом есть разные мнения **"как разбить / выделить фичу"**
> Как минимум - следующие:

1. По тому, как они выглядят и примерно работают
   *`posts-list`, `post-details`*

   > При этом фичи "разбухают", и есть ~~камень в огород~~ претензия, что они "не несут бизнес-ценность"

2.  По названию модели
   *`posts`*

    > Во многих случаях без такой интерпретации никуда, но опять же - есть риск разбухнуть, плюс поломать абстракции (особенно когда у нас будет еще папка `entitites`)

3. Одна фича разбивается по другим абстракциям (чтобы как раз нести "бизнес-ценность" и быть достаточно маленькой)
   *`features/.../models`, `entities`*

   > Иногда кажется логичным, но с таким подходом (лично у меня), возникает кучу вопросов, в том числе вы (@sergeysova , @karina-drummer ) [сами подмечаете](https://github.com/feature-sliced/wiki/discussions/18#discussioncomment-422752), что в таком случае логика по `features/../models/` и `entities` размазывается, и без переименования (`models` => `entities`) остается много вопросов ~~(хотя как по мне вопросов становиться только больше :D)~~

---

### 📣 Приглашаю к ~~холивару~~ дискуссии!

<details>
<summary>В прошлый раз на этом остановились</summary>

![image](https://user-images.githubusercontent.com/42924400/110217128-774a1b80-7ec3-11eb-9c30-417d36020adc.png)
</details>

**Comment by azinit** [comment]
![image](https://user-images.githubusercontent.com/42924400/110348271-86a3a300-8042-11eb-865d-ab43bcf8a3a9.png)

**Reply by azinit** [reply]
> ## Сначала пооффтоплю, а потом к твоей мысли))

Вот тут как раз и опасная фразочка)

> "Фича внутри фичи? Да"

Обычно это значит что у нас 
- либо **одна оч комплексная одна фича**:
   ```sh
   - features/auth-form/oauth-google
   - features/auth-form/by-phone
   ```
    > Хотя это еще неплохо, если `auth-form`, как говорил @sergeysova - чисто для "группировки фичей", но не более

- либо функциональность, *корректно* (каждый определяет по-своему) разбитая по абстракциям
   ```sh
   - features/oauth-google
   - features/by-phone
   - pages/auth/form
   - processes/auth
   ```
   > В некоторых случаях (если причесать пример), как будто это тоже уместно может быть

---

Кажется логичным на самом деле, делать лучше группировку фичей по смыслу (и заявить об этом в методологии возможно), но осторожно

### Тогда у нас выходит `auth-form` - "группа фичей (но не фича сама по себе))", а `oauth-google`, `by-phone` - "самостоятельные фичи"
> Но опять же - холиварно, и как будто это тоже надо оставлять на откуп архитектору - хотя это базис для методологии

---

> UPD: Второй пример плох еще тем, что у нас функциональность авторизационной формы - расплывается снова по проекту, и это опять нас возвращает ~~к каменному веку~~ ко времени разделения "по типам"

**Reply by azinit** [reply]
А теперь к тезисам основным:

> сама по себе форма фичей не является. чтобы она ей стала — нужен некий контекст, например, аутентификация

**Соглашусь, звучит логично**
  > Особенно, если мы рассматриваем альтернативу
  > ```sh
  > - features/oauth-google
  > - features/by-phone           # или features/auth-by-phone
  > ```

--- 

> надо ли её держать отдельно? нет.
она попадает в контекст аутентификации

Видимо тут ты как раз про "группирующую папку", и про "контекст использования", т.е. сам процесс `processes/auth`
Вродь тож вписывается хорошо)

---

> надо ли рядом с отображением держать идею? нет.
идея не знает об образе.

Под идеей ты тут подразумеваешь "как будет использоваться это форма авторизации"? Или сам процесс "авторизации"?

Т.е. я понимаю, что в принципе группа фичей `features/auth-form` в принципе не должна знать где использоваться, да

Но былоб супер, если бы ты тезис последний развернул)

**Reply by binjospookie** [reply]
Попробую ответить на всё сразу)

Идея бизнеса в данном случае = нужна аутентификация. Её можно обозвать `auth`. В моём мире это был бы пакет.
У которого есть 3 директории:
* integration (как запустить фичу)
* model (логическая часть)
* UI (😺)

## Внутри пакета
### Integration
В идеале -- экспортируемая функция makeAuthFeature. Которая принимает переводы кнопок, наполняет модель данными, прикидывает транспорт итп.

### Model
Отображение бизнес-идеи без UI. Из доменов, которые сразу же можно выделить:
* phone
* email
* oauth-google

Каждый домен отвечает за какую-то часть модели.

Если бы появился vk oauth, то структуру я бы сделал такой:
* phone
* email 
* oauth/google
* oauth/vk

На верхнем уровне видно 3 сущности. А суммарно их 5. А в oauth всего 2. Такое разбиение позволяет тебе погрузиться в нужный контекст плавно (на мой личный взгляд, офк)

### UI
Здесь просто находятся "шаблоны". Например, у нас есть 2 вида отображения формы аутентификации:
* template1 
* template2

*названия условны*

Каждый из этих шаблонов "цепляется" к публичной части модели. Сама модель ничего не знает про этот слой (идея не знает об образе).

## Итого
На основе идеи "нужна аутентификация" мы создали пакет, в котором выделили 3 слоя: 
1) голая идея осталась в model
2) определили внешние виды
3) дали возможность интегрировать фичу, скрыв остальные слои

Как-то так)

**Reply by azinit** [reply]
@binjospookie спасибо, что развернул!)
На самом деле, оч классно расписал и некоторые аналогии можно словить с этого всего

> Integration: В идеале -- экспортируемая функция makeAuthFeature, которая ... [принимает кастомизацию]

Оч похоже на Public API которая каждая абстракция (shared/entity/feature/process) должна экспортировать, которым можно пользоваться при ее инициализации

*Т.е.  здесь "идея" может проявляться и как фича, и как процесс*

---

> Model: Отображение бизнес-идеи без UI. Из доменов, которые сразу же можно выделить

А это оч напоминает `feature/{foo}/store` (@sergeysova , @karina-drummer  не бейте)

*Т.е. логика фичи в виде самих данных / стейта / редьюсеров / thunks / мидлваров / эффектов и т.д. - aka data-layer, без приявзки к отображению*

---

> UI: Здесь просто находятся "шаблоны". Например, у нас есть 2 вида отображения формы аутентификации [...] и каждый из этих шаблонов "цепляется" к публичной части модели

А это (лично у меня) обычно находится в `feature/{foo}/components`

*Т.е. его компоненты и UI, который, в том числе, может представлять собой и несколько экспортируемых компонентов - которые так или иначе цепляются к "идее" - т.е. store (но не лезут глубоко внутрь, а обращаются к публичному API `"../store"`)*

*И действительно - store (или model у @sergeysova , @karina-drummer ) понятия не имеет как он будет тягаться - внутри фичи, или на уровне страницы (ведь на то он и публичный) - например, для построения реактивности между фичами*

---
---

> На основе идеи "нужна аутентификация" мы создали пакет, в котором выделили 3 слоя:
> - голая идея осталась в model
> - определили внешние виды
> - дали возможность интегрировать фичу, скрыв остальные слои

И по итогу да, верно, получили что-то вроде (в моем понимании):

```ts
└── features/                            # Или это может быть даже process / entity
    └── auth/
            ├── components/              # "Внешний вид" или UI нашей идеи (ее образ)
            ├── store/                   # Сама "идея" функциональности
            └── index.ts                 # Публичный API фичи/процесса/(пакета)/... - в том числе для ее интеграции и использования
```

**Reply by binjospookie** [reply]
Всё верно, да

Integration может быть директорией или файлом. Зависит от того, на сколько там много строк, по сути)

**Reply by azinit** [reply]
И на самом деле все что ты тут расписал @binjospookie - оч хорошо показывает, что у нас есть

- High-level abstractions: `app`, `processes`, `pages`, `features`, `entities`, `shared`
   > UPD: Перечислены от наиболее "знающих" к наименее: `page` знает про `features`, но не про `process` где участвует и т.п.
- Middle-level abstractions: `ui` | `components`, `lib` (aka helpers-lib), `model` | `store`, `api`

**И почему-то лично мне кажется логичным задекларировать в спецификации помимо этого следующее:**
1. **Каждая high-level абстракция** располагается только на высшем уровне, один раз (но не внутри друг друга)
   ```diff
   - features/foo/entities
   - features/foo/share
   + features/foo/

   - entities/bar/shared
   + entities/bar/
   ```
2. **Каждая middle-level абстракция** располагается **только** внутри high-level (но не на высшем уровне)
   ```diff
   - api/posts/
   + shared/api/posts

   - ui/button
   + shared/ui/button
   + features/foo/components    # Да будет холивар как именовать ui
   + features/foo/ui

   - lib/dates
   + shared/lib/dates
   ```

3. **Каждая абстракция** должна быть спроектирована **с публичным API для использования и интеграции**
   ```diff
   - import { actions } from "features/auth/store/slice"      # Для dispatch(actions) (да я редаксоид)
   + import { authStore } from "features/auth"  # dispatch(authStore.actions)

   - import { formatDate } from "shared/lib/dates/format"
   + import dates from "shared/lib/dates"

   - import { ButtonGroup } from "shared/button/group"
   + import { ButtonGroup } from "shared/button"    # Или Button.Group (привет Compound.Components)
   ```

4. Желательно максимально разделять образ (отображение/ui) и идею (store/model)
   ```diff
   # Плохо располагать модель и отображение на одном уровне в абстракции
   - features/auth/actions.ts
   - features/auth/form.tsx
   # По @sergeysova , @karina-drummer 
   + features/auth/models/actions  # идея
   + features/auth/ui/form.tsx # UI
   # По @martis-git , @niyazm524 
   + features/auth/store/actions  # идея
   + features/auth/components/form.tsx # UI

   - entities/user/actions
   - entities/user/models
   + entities/user/components
   + entities/user/ui
   ```

**Reply by azinit** [reply]
@binjospookie @ilyaagarkov @sergeysova @karina-drummer @AlexandrHoroshih @niyazm524  приглашаю к оценке того - что выписал тут на основании доводов @binjospookie и наших обсуждений

Если все гуд - то можно даже в доку куда-нить вынести (мб даж в ту самую схему внести, которую под конец итерации хотели показать)

**Reply by binjospookie** [reply]
Конкретно по: `app`, `processes`, `pages`, `shared` ничего вразумительного сказать не смогу (по крайней мере, на данный момент), тк в моём мире они отсутствуют.

Но предполагаю, что `shared` я называю `lib`

**Reply by azinit** [reply]
Еще мне понравилась наша мысль с обсуждения что "Фича определяется в момент использования, как и контекст функции в JS"

*Тоже мб получится как-нибудь впихнуть (хотя по идее на этом все базируется - то что страница/фича/... - не знает как она будет юзаться*

**Reply by binjospookie** [reply]
Есть мысль, но она очень сырая: каррирование фичи

**Reply by ilyaagarkov** [reply]
мне не очень нравится папка `shared`
не вижу никаких проблем чтобы разместить ui/lib (кстати почему lib a не libs?) /api на верхнем уровне.

Так как в итоге у нас все равно будут строгие правила, кто и от куда может импортировать, то смысл делать папку shared?

Мое мнение, чем быстрее я могу добраться до любого элемента тем удобнее. 

Пока не могу придумать пример, когда в фиче должна быть папка `lib`. Все таки `lib` это код не связанный  с бизнесом, тогда как фича должна содержать только бизнесовый код

**Reply by ilyaagarkov** [reply]
> Еще мне понравилась наша мысль с обсуждения что "Фича определяется в момент использования, как и контекст функции в JS"

Звучит красиво, но есть ощущение, что это будет избыточно для большенства фич.
То есть это можно оставить как рекомендацию для случаев когда это нужно

Хотя может я просто мысль не до конца понял

**Reply by binjospookie** [reply]
> Пока не могу придумать пример, когда в фиче должна быть папка lib
Плюсик. 

Lib это что-то, что помогает строить entties, features, process (наверное). Что можно причислить к lib:
* ui-kit
* helpers (Ramda like)
* либу форматирования валют
* и подобные

Либы не должны зависеть друг от друга, но могут встречаться исключения.

**Reply by azinit** [reply]
@ilyaagarkov чую в голосовом чате в feature-sliced было бы быстрее ахах

**Reply by azinit** [reply]
@ilyaagarkov 

> мне не очень нравится папка shared
не вижу никаких проблем чтобы разместить ui/lib (кстати почему lib a не libs?) /api на верхнем уровне.

Там как раз выше и описал почему)

- `shared` - у нас high-level абстракция
- `ui`, `lib(s)`, `api`, `store` - это middle-level абстракции, которые могут быть подмодулями/подабстракциями high-level (выше пример)

На вопрос "почему мы не можем вынести `ui`/`libs` наверх" - соответственно, ответ 

**"Т.к. если разраб будет на одном уровне держать и features и ui/libs - то мы возвращаемся к другим итерациям методологии - когда со стороны сложно было понять - а что-где располагать?"**

**Если же человек видит явно "Есть HIGH-абстракции, которые он всегда юзал, а есть привычные ему MIDDLE-абстракции, которые теперь более вложенно и структурированно располагаются - вопросов (как по мне) - гораздо меньше"**

---

**Reply by azinit** [reply]
@ilyaagarkov 

> Так как в итоге у нас все равно будут строгие правила, кто и от куда может импортировать, то смысл делать папку shared?

Вот в этом и соль - как ты поймешь что **"здесь" можно располагать `ui/lib`, а "тут" нельзя?**

**Это решение (как по мне, и как мы с ребятами обсуждали), добавляет интуитивности и понятности разрабу, и т.к. эти правила можно легко формализировать (более легко чем `ui/lib` могут быть на самом верху, или же внутри high-level абстракций) - то получается более простая и гибкая методология (что упрощает и разработку линтера и понимание со стороны разрабов)**

**Reply by azinit** [reply]
@ilyaagarkov 
> Мое мнение, чем быстрее я могу добраться до любого элемента тем удобнее.

Спорили об этом и с @sergeysova , но вроде сошлись на том, что:
1. Если мы судим о методологии только с точки зрения *"Хочу меньшей вложенности, давайте все абстракции ui/lib/api вынесем наверх!"*
    - Как вариант - добавить альясы для либ (но не приветствуется, т.к. опять же путает связи между абстракциями)
    - Как отчаянный  вариант - готе вернемся к структуризации по типам, когда есть большие папки `components/`, `store/`, `api` к которым сразу есть доступ - но зато из-за этого оч распыляется логика - из-за чего мы как раз и пришли все дружно к идеи разделении по фичам

2. Иначе мы смиряемся с тем, что можем воспользоваться излишней вложенностью, но можем направить ее на большую интуитивность
    - Вот у тебя есть два стула:
       - Структура 1: `app`, `api`, `lib`, `ui`, `features`, `pages`
       - Структура 2: `app`, `pages`, `features`, `shared` (и внутри каждого уже может быть разделение по типам: ui, lib, api)
       
       > На какой сядешь сам и команду посадишь? (и чтобы у них не было вопросов вроде "А поч я не могу создать `lib/features`?")

**Reply by azinit** [reply]
@ilyaagarkov на самом деле спасибо, я тут пока объяснял вспомнил еще один момент:

High/middle абстракции можно понимать следующим образом:
- `High-level-abstraction`: абстракция "по функциональности" (методология сначала разбивает по ним)
- `Middle-level-abstraction`: абстракция "по типам" (то как делили раньше, но теперь у нас каждая абстракция делится по типам, и вроде выглядит логично)

> UPD: [Здесь](https://github.com/feature-sliced/wiki/discussions/31#discussioncomment-451507) и [здесь](https://github.com/feature-sliced/wiki/discussions/31#discussioncomment-451612) получше развернул

**Reply by azinit** [reply]
@ilyaagarkov 

> Пока не могу придумать пример, когда в фиче должна быть папка lib. Все таки lib это код не связанный с бизнесом, тогда как фича должна содержать только бизнесовый код

1. Ты говоришь скорее про `shared/lib` (т.е. т.к. он shared - он не связан с БЛ почти, но т.к. lib - это инфраструктурные решения/хелперы)
2. А если мы в целом относимся к `lib` по типу - то это ведь почти структурированный набор хелперов, верно? (это как отличать `shared/ui` и `features/foo/ui`)
   - И в зависимости от абстракции где находится - решает задачи только для своего типа (не лезет в ui например), но в соответствии со своей абстракцией-родителем 
    > - т.е. либо общие хелперы для проекта - (`shared/lib`)
    > - либо хелперы для этой entity (`entities/lib`)

**Reply by azinit** [reply]
@ilyaagarkov 

Да, само собой - не всегда у тебя в фиче будет `lib`, но мы предлагаем так делить (по крайней мере я с @niyazm524 ), поскольку вчерашний человек который сначала делил по `store/users` , сможет сейчас делить наоборот по `**/users/store`

> UPD: Т.е. это понятно и для новых людей, и понятно тебе как разрабу (да тебе мб проще будет импортить ui/lib если они будут на самом верху - но не будет ли это во вред методологии?)

**Reply by azinit** [reply]
> > @martis-git
> > Еще мне понравилась наша мысль с обсуждения что "Фича определяется в момент использования, как и контекст функции в JS"
>
> @ilyaagarkov 
> 
> Звучит красиво, но есть ощущение, что это будет избыточно для большенства фич.
> То есть это можно оставить как рекомендацию для случаев когда это нужно
> 
> Хотя может я просто мысль не до конца понял

Согласен, мы тоже пока не до конца понимаем)

В большинстве случаев это мб и рили будет "ту мач", но я пока вижу две интерпретации:
- *"У тебя есть фича (или другая абстракция), и ее польза и ценность будет определена в момент использования (на странице/процессе - или в целом более высокой абстракции)"*
- *"У тебя есть абстракции-типы `components`, `lib`, `api`. Но их применение, цели и польза будут определяться в зависимости от абстракции-функциональности, в которой она находится (Яркий пример: `shared/components` и `features/{foo}/components`)"*

> Это к вопросу опять - почему в методологии должно быть слово `feature-***`

**Reply by azinit** [reply]
@ilyaagarkov но спасибо за замечания, т.к. они позволили хорошо сформулировать мне некоторые вещи)

> Обсудим позже с нашим комитетом, и [в целом](https://github.com/feature-sliced/wiki/discussions/31#discussioncomment-451507)

**Reply by azinit** [reply]
> Lib это что-то, что помогает строить entties, features, process (наверное). Что можно причислить к lib:
> 
> * ui-kit
> * helpers (Ramda like)
> * либу форматирования валют
> * и подобные
> 
> Либы не должны зависеть друг от друга, но могут встречаться исключения.

@binjospookie тож приглашаю к [некст итерации дискуссии](https://github.com/feature-sliced/wiki/discussions/31))

**Comment by azinit** [comment]
https://github.com/feature-sliced/wiki/discussions/18#discussioncomment-450669

![image](https://user-images.githubusercontent.com/42924400/110350830-2bbf7b00-8045-11eb-8795-aeee057984f8.png)

**Reply by binjospookie** [reply]
про entities прям хорошо. могу лишь додать пример: form. некая сущность, объединяющая state-manager (ff, например) и UI-kit. Позволяющая строить любую форму на основе входящих данных (json с полями)

Чтобы такая сущность стала фичей -- надо додать данных. На основе контекста данных (это про аутентификацию, фидбек, оплату) определяется контекст фичи.

**Reply by ilyaagarkov** [reply]
мне все таки кажется что entities это не про абстрактные понятие как form или filters.

Как я вижу развитие всех этой движухи, `entities` отпочковались именно от` features`, в результате того что фичи наподобие  `features/post` приводили  к необходимости импортов одних фич в другие. То есть какая то часть того что считалось `features` перешли в `entities`.
Но как и `features`, `entities` все еще должны быть максимально близко к бизнесу.

**Reply by azinit** [reply]
@ilyaagarkov

> Как я вижу развитие всех этой движухи, entities отпочковались именно от features, в результате того что фичи наподобие features/post приводили к необходимости импортов одних фич в другие.

Соглашусь, т.к. до этого 
- либо создавали оркестрацию фич для их кросс-коммуникации (кластеры или на уровне страницы)
- либо выносили в третью фичу (которую было незашкварно заимпортить)

Сейчас же ребята (@sergeysova , @karina-drummer ) предлагают для этого юзать именно `entities` 
*(где будут более оторванные от фичи абстракции, благодаря которым можно разрабатывать изолированные фичи)*

> мне все таки кажется что entities это не про абстрактные понятие как form или filters.
> Но как и features, entities все еще должны быть максимально близко к бизнесу.

**Да, и по-моему и я там написал - что `entities` - это в том числе про БЛ)**
*Максимально абстрактные у нас - это `shared`-товарищи*

Просто надо (как по мне) понимать разницу и суть:
```sh
- features/tasks-filtered-list # юзает `filtered-list`
- features/items-filtered-list # юзает `filtered-list`
- entities/filtered-list # не знает про tasks/items, но может быть адаптирован под использование для них
```

**Comment by ilyaagarkov** [comment]
Накину своих мыслей

Я не могу дать четкое определние термину фича, но я попытаюсь ответить на вопрос что такое список фич проекта (upd: в результате все таки пришел к определению =) ).

Список фич проекта это тот набор фич который бы перечислил бизнес, пытаясь описать возможности своего продукта конечному пользователю(в целом это мое понимание слов "код, который несёт бизнес ценность").

При этом если бизнес говорит пользователю "Хей смотри у нас  **_есть_** стена" - это не фича. Если у нас что-то "**_есть_**", то это что-то - entity

Если бизнес говорит "смотри ты можешь **_писать_** посты на стене" - **_написать_** пост на стене это фича. Действие которое может выполнить пользователь и есть фича.
Тут сразу надо оговориться, что не любое дейсвие которое можеть выполнить пользователь является фичей. Пользователь может перейти на какую то страницу, но это явно не фича. Так что нужно явно выделить какие то критерии

Например
- действия пользователя должны привести к изменению в бд (или любого другого хранилища)
- пользователь должен явно видеть результаты этого изменения (сбор статистики посещеных страниц таким образом не попадает в фичи - хоть в результате действий пользователя и меняется бд, пользователь этого явно не видит) _Данный пример спорный, возможно что сбор статистики это все таки фича =)

Фича содержит в себе информацию о том, 
- какие данные нужны для того-го или иного действия (как пример - входящие параметры функциии)
- как валидировать данные
- где взять недостобщие данные (взять что-то из общего state)
- досупна ли данная фича (на пример может ли текущий пользователь писать посты на стене другого пользователя)
- какие данные и как нужно их изменить при действии пользователя

а так же фича может содержать
- компоненты специфичные для конктретной фичи. (Например форма отправки поста или даже какие-то менее значимые компоненты но уникальные для данной фичи)

**Reply by azinit** [reply]
Хорошая простыня текста вышла)

**Reply by azinit** [reply]
Ох, на самом деле хорошо вещи расписал с философской стороны (Сущность ~ `entities`, Действие - `features` (но есть критерии)

Завтра наверн подробнее гляну еще (т.к. есть что дополнить и обсудить)

Но пока также приглашаю к следующей итерации обсуждения - ["Абстракции методологии, их цели и нейминг"](https://github.com/feature-sliced/wiki/discussions/31))

**Reply by sergeysova** [reply]
Отлично описано. Разве что "сбор статистики" это не фича для потребителя продукта, ведь это детали реализации, как именно она будет сделана — не важно. Важно лишь то, как пользователь будет использовать результаты сбора.
Ведь пока он не может их использовать, для него этого нет.

**Comment by sergeysova** [comment]
Можно отсюда вытащить немного определений #43
